types/node/globals.d.ts

```diff
@@ -197,18 +197,18 @@ declare namespace NodeJS {
         syscall?: string | undefined;
     }
 
-    interface ReadableStream extends EventEmitter {
+    interface ReadableStream<R = any> extends EventEmitter {
         readable: boolean;
-        read(size?: number): string | Buffer;
+        read(size?: number): R | null;
```

Comment:
This is incorrect. The output of `read` is influenced by:

1. the stream being in object mode or binary mode
2. `setEncoding()` being called or not

On top of that, I'm not sure what the current typing is, but this change will likely cause breakage.

Reply:
Your points are correct but I believe this PR already addresses 2 of them:

> 1. the stream being in object mode or binary mode

Given that `objectMode` is set when the stream is constructed, this is covered by setting the generic to be `Readable<{ foo: string, bar: number }>` etc for an "object mode" stream vs `Readable<string | Buffer>` for a "regular" stream. It _is_ possible to set the generic type to not match the `objectMode `, but to enforce this with types would really complicate the constructor arg types. Ultimately, it is up to the producer of the streams to correctly set the generic type in order for the consumer to benefit.

> 2. setEncoding() being called or not

Given a "regular" binary mode stream, the type is `Readable<string | Buffer>`. This is the type that is already used in the fs readable streams (e.g `on(event: 'data', listener: (chunk: Buffer | string) => void): this`). The existing types handle `setEncoding()` by using a union of the possible modes.

> On top of that, I'm not sure what the current typing is, but this change will likely cause breakage.

There are two potential breakages that you might be talking about here:

1. *New generic type argument (`Readable` => `Readable<T>`)*. I've set a default of `T = any` similar to the `stream/web` types. This _does_ mean that `$ExpectType` sees a difference but it shouldn't be a breaking change. Existing implementors of the `NodeJS.ReadableStream` _interface_ will see their types being less refined though, as `read(size?: number): string | Buffer` becomes `read(size?: number): any` due to the generic not being set. Existing extenders of the `Readable` _class_ will see no change as the types are already set to `any`.

2. *`read(size?: number): any` => `read(size?: number): string | Buffer | null` for fs streams*. This *is* a breaking change. I can remove the `| null` (https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/60526 to match the [spec](https://nodejs.org/dist/latest-v18.x/docs/api/stream.html#readablereadsize)) but that still leaves `read(size?: number): any` => `read(size?: number): string | Buffer` which is still a breaking change...

Reply:
The `| null` one for `read()` is needed. This is not just for fs streams but for all streams: `.read()` will return `null` on occasion.

Let's recap. If the stream is in object mode, `read()` must be typed as `T | null`. If the stream is in binary mode, the chunk could be of type `Buffer | null`. If `.setEncoding('utf8')` (or others encoding) has been called, the type could be of type `string | null`.

Reply:
> The | null one for read() is needed. This is not just for fs streams but for all streams: .read() will return null on occasion.

Cool, I'm totally happy with keeping that breaking change (number 2) in.

> Let's recap. If the stream is in object mode, read() must be typed as T | null. If the stream is in binary mode, the chunk could be of type Buffer | null. If .setEncoding('utf8') (or others encoding) has been called, the type could be of type string | null.

yep, given that the object mode flag is set at when the string is constructed (and the docs say ["Attempting to switch an existing stream into object mode is not safe"](https://nodejs.org/api/stream.html#object-mode)), We can safely say that setting the generic `T` to whatever the object type is for an object stream should be all correct in this PR, which I think we agree on?

With binary mode, we can't decide if `read()` will return `Buffer | null` or `string | null` because `.setEncoding(...)` could be called at any time. The best we can do with is combining the two into `Buffer | string | null`.